ENH: Add lychee link-checker as a Utilities/Maintenance script

jhlegarreta opened the pull request:
Add a manual lychee link-checker (`Utilities/Maintenance/check-links.sh` + `lychee.toml`) for periodic maintainer-driven runs, instead of the per-PR GHA workflow originally proposed.

<details>
<summary>Why a manual script instead of a GHA workflow</summary>

The original workflow attempt failed CI consistently due to:
- HTTP 429 rate-limits from github.com on commit URLs in release notes (no per-run workaround that doesn't also accept real broken links).
- [lycheeverse/lychee#1574](https://github.com/lycheeverse/lychee/issues/1574) — caching across CI runs is not effective.
- False positives on host-specific bot blocks (999 from LinkedIn, etc.).

In the [Feb 2026 discussion](https://github.com/InsightSoftwareConsortium/ITK/pull/5377#issuecomment-2628200000) on this PR, @jhlegarreta agreed the per-PR CI approach was not workable and authorized a pivot. This commit takes over the branch (with `Co-Authored-By:` credit) and converts to the script approach.

</details>

<details>
<summary>What's included</summary>

| Path | Purpose |
|---|---|
| `Utilities/Maintenance/check-links.sh` | Bash wrapper. Resolves the repo root, requires `lychee` on `PATH`, accepts optional path arguments, writes a Markdown report and persistent cache. |
| `Utilities/Maintenance/lychee.toml` | Configuration with rate-limit-aware `accept = [..., 429, 999]`, commit-URL skip regex, ThirdParty exclusion, and document-only `include` globs. |
| `.gitignore` | Adds `.lycheecache` and `.lychee-report.md` (local artifacts). |

Intended use: a maintainer runs `Utilities/Maintenance/check-links.sh` periodically (or scopes it to a subdirectory) and acts on the report. No CI gate is added.

</details>

<!--
provenance: claude-code session 2026-05-09
related_files:
  Utilities/Maintenance/check-links.sh
  Utilities/Maintenance/lychee.toml
takeover_authorized_at: https://github.com/InsightSoftwareConsortium/ITK/pull/5377#issuecomment-2628200000
original_author: jhlegarreta (Co-Authored-By trailer on the commit)
post_merge_action: optionally schedule a recurring run via maintainer cron / Discourse callout
-->

dzenanz commented:
Many failures are due to HTTP code 429: Network error: Too Many Requests. Can we slow down request rate or something similar?

jhlegarreta commented:
Looks like the throttling does not work. Have gone through these
https://github.com/lycheeverse/lychee-action?tab=readme-ov-file#utilising-the-cache-feature
https://github.com/lycheeverse/lychee/blob/master/docs/TROUBLESHOOTING.md

but have not found how to modify the arguments to make it work. Have seen other workflows accepting 429, e.g.
```
--accept 403,429,500,502,999
```

but I guess that will effectively not check the links at issue.

Also, checks will not work until this issue is solved:
https://github.com/lycheeverse/lychee/issues/1574

I do not know any other tool that does this job for rst files. This one only seems to work for md files:
https://github.com/tcort/markdown-link-check

jhlegarreta commented:
@hjmjohnson thanks for the heads-up; going through challenging times on my end, so I am being unable to push ITK items. My sincere apologies. The approach you propose sounds reasonable. Feel free to close the PR.

hjmjohnson committed:
ENH: Add lychee link-checker maintenance script

Replaces the original GHA workflow attempt (PR #5377) with a manual
script under Utilities/Maintenance/.  Per author + maintainer
discussion on the original PR, integrating lychee into per-PR CI is
not workable: lycheeverse/lychee#1574 and HTTP 429 rate-limits on
github.com / DOI hosts produce too many spurious failures to act on
in CI.

New artifacts:
  - Utilities/Maintenance/check-links.sh — wrapper that resolves the
    repository root, requires lychee on PATH, runs against the supplied
    paths (or the whole tree by default), and writes a Markdown report
    plus a persistent cache.
  - Utilities/Maintenance/lychee.toml — configuration with the rate-
    limit-aware accept list (treats 429/999 as non-broken), commit-URL
    skip pattern that motivated the original CI failures, exclusions
    for ThirdParty trees, and path globs limited to documentation
    file types.
  - .gitignore: ignore the local cache (.lycheecache) and report
    (.lychee-report.md) artifacts so re-running does not pollute the
    working tree.

This script is intended for periodic / on-demand runs by maintainers,
not the per-PR pipeline.

Co-Authored-By: Jon Haitz Legarreta Gorroño <5576557+jhlegarreta@users.noreply.github.com>

hjmjohnson commented:
@jhlegarreta — thank you for the original work and for the OK to pivot. Force-pushed `cc6bfedde3` to `AddLinkCheckerGHAWorkflow` (lease pinned to your prior tip). The branch now contains a manual `Utilities/Maintenance/check-links.sh` + `lychee.toml` instead of the GHA workflow, with you credited via `Co-Authored-By:` on the commit. Title and body updated to match.

Holler if you'd rather I close this and open the script under a fresh PR — both work; reusing this branch felt cleaner since the conversation history is already here.

hjmjohnson committed:
DOC: Refresh four obvious link-rot targets in user-facing docs

Updates four URLs flagged by Utilities/Maintenance/check-links.sh that
have well-defined modern equivalents:

- Doxygen `\tparam` reference: stack.nl/~dimitri/doxygen ->
  doxygen.nl/manual (the canonical Doxygen documentation site since
  the project's move).
- SCI Institute Seg3D landing page: cibc-software/seg3d.html ->
  sci.utah.edu/seg3d.
- SCI Institute SCIRun landing page: cibc-software/scirun.html ->
  sci.utah.edu/scirun.
- MITK home: mitk.org/wiki/The_Medical_Imaging_Interaction_Toolkit
  (the wiki page no longer resolves) -> mitk.org/ (the project home).

All four replacements were verified to return HTTP 200.

Other lychee findings (release-notes link rot, ipfs.io connection
resets, Kitware blog post moves) need case-by-case research and are
deferred.

hjmjohnson committed:
DOC: Modernize ITK 5.x doxygen URLs and migration-guide path

Bulk update of broken links flagged by
Utilities/Maintenance/check-links.sh on top of the prior obvious-fixes
commit:

- itk.org/Insight/Doxygen/html/...
    -> docs.itk.org/projects/doxygen/en/stable/...
  (the legacy doxygen path moved to docs.itk.org).

- classitk_1_1Experimental_1_1<Range>.html
    -> classitk_1_1<Range>.html
  for ImageBufferRange, IndexRange, ShapedImageNeighborhoodRange,
  ImageRegionRange.  These classes were promoted out of the
  itk::Experimental namespace; only the un-Experimental URL resolves
  on the modern doxygen build.  HyperrectangularImageNeighborhoodShape
  is touched the same way (the un-Experimental name is the live one).

- namespaceitk_1_1Experimental.html -> namespaceitk.html
  (the namespace was retired; its members live in itk:: now).

- github.com/.../ITK/blob/master/Documentation/ITK5MigrationGuide.md
    -> github.com/.../ITK/blob/main/Documentation/docs/migration_guides/itk_5_migration_guide.md
  (file moved + renamed; anchors in the old URL resolve in the new
  layout).

- Documentation/Maintenance/Release.md: ReleaseNotes folder rename
  -> Documentation/docs/releases.

- courses.md: uu.nl/en/master/medical-imaging/study-programme
    -> uu.nl/en/masters/medical-imaging.

All replacement URLs verified to return HTTP 200 before staging.
Re-running check-links.sh on the touched files reduces error count
from 90 to 17 (residual = dead course pages, ipfs.io infra, HDF5
license relocation, opencollective rate-limit transient — all need
case-by-case research).

hjmjohnson committed:
DOC: Repoint 5.0b03 release-note links to current main-branch paths

Six links in Documentation/docs/releases/5.0b03.md pointed at files
that have moved or been renamed since the 5.0b03 release.  Repoint
each to its current location on main:

- .github/ISSUE_TEMPLATE.md    -> .github/ISSUE_TEMPLATE/ (templates
                                  moved to a directory of files).
- .github/PULL_REQUEST_TEMPLATE.md
                               -> .github/pull_request_template.md
                                  (renamed lowercase).
- Documentation/CodeOfConduct/Motivation.md
                               -> CODE_OF_CONDUCT.md (the standalone
                                  Motivation.md was merged into the
                                  top-level Code of Conduct).
- Documentation/Data.md        -> Documentation/docs/contributing/data.md
                                  (docs reorganised under
                                  Documentation/docs/).
- Documentation/UploadBinaryData.md
                               -> Documentation/docs/contributing/upload_binary_data.md.
- Utilities/UploadBinaryData.sh (script removed)
                               -> Documentation/docs/contributing/upload_binary_data.md
                                  (the canonical doc that supersedes
                                  the now-removed helper script).
- Documentation/ReleaseNotes/  -> Documentation/docs/releases/.

Replacement URLs verified to return HTTP 200.  Two remaining 404s in
this file (atlassian.net JIRA project — decommissioned, the broken
link itself communicates that fact; GitCheatSheet.pdf — file removed
without a successor) are left in place for historical accuracy.

hjmjohnson committed:
DOC: Drop the dead Atlassian JIRA link in 5.0b03 release notes

The legacy ITK JIRA project at insightsoftwareconsortium.atlassian.net
was decommissioned by Atlassian; the URL no longer resolves.  Remove
the broken link and rephrase the surrounding sentence to state the
decommissioning directly.

hjmjohnson committed:
DOC: Drop three retired course bullets from learn/courses.md

The University of Central Florida (cs.ucf.edu/~bagci/teaching/mic17),
Uppsala (it.uu.se/edu/course/homepage/bild1/vt14), and Western
University (eng.uwo.ca/biomed/courses/courses_9519) course pages have
been retired with no announced successor URLs.  Remove the three
bullets rather than carry permanently broken links.

hjmjohnson committed:
DOC: Repoint the GDCM Copyright link to the GitHub repository

gdcm.sourceforge.net/Copyright.html no longer resolves; GDCM
development moved to github.com/malaterre/GDCM, where the canonical
copyright file is Copyright.txt at the repo root.

hjmjohnson committed:
DOC: Drop the ITKv5Preparation section from the ITK 5 migration guide

The Utilities/ITKv5Preparation directory contained one-shot bash
scripts used during the 4 -> 5 migration; the directory was removed
once the migration completed.  The "Update scripts" section pointed at
that now-deleted directory and trailed off mid-sentence; remove it
rather than carry a permanently broken link to scripts that no longer
exist.

hjmjohnson committed:
DOC: Fix the empty itk::TIFFImageIO link in learn/faq.md

The TIFF row of the supported-formats table had a markdown link with
an empty URL ([\`itk::TIFFImageIO\`]()).  Repoint to the actual
TIFFImageIO doxygen page on docs.itk.org.

hjmjohnson committed:
DOC: Repoint the HDF5 LBNL license link to the consolidated LICENSE file

HDF Group reorganised the HDF5 license layout: COPYING_LBNL_HDF5 was
renamed to LICENSE_LBNL_HDF5 and then consolidated into the single
top-level LICENSE file at the repository root.  The legacy
support.hdfgroup.org/ftp/HDF5/releases/COPYING_LBNL_HDF5 URL no longer
resolves.

Update the licenses.md note to point at
https://github.com/HDFGroup/hdf5/blob/develop/LICENSE, which contains
the LBNL Copyright Notice and Licensing Terms verbatim.

hjmjohnson committed:
DOC: Repoint the HDF5 site link in faq.md to hdfgroup.org/solutions/hdf5

The HDF Group retired the support.hdfgroup.org/HDF5/ landing page; the
canonical HDF5 home is now www.hdfgroup.org/solutions/hdf5/.

hjmjohnson committed:
DOC: Replace the dead itk.org/CourseWare PDF with the Software Guide

itk.org/CourseWare/Training/RegistrationMethodsOverview.pdf no longer
exists.  Repoint the "registration overview" sentence in faq.md to the
Registration chapter of the ITK Software Guide (Book 2, Chapter 3),
which is the canonical successor and is actively maintained.

hjmjohnson committed:
DOC: Drop the dead HyperrectangularImageNeighborhoodShape link in 5.0b01

The Hyperrectangular shape no longer has its own doxygen page on the
modern docs.itk.org build; demoting that markdown link to inline code
keeps the class name visible without pointing at a 404.  The sibling
ShapedImageNeighborhoodRange page does still exist and remains linked.

hjmjohnson committed:
DOC: Drop the dead GitCheatSheet.pdf reference in 5.0b03

Documentation/GitCheatSheet.pdf was removed from the repository
without a successor.  Remove the trailing "We also have a Git
cheatsheet for quick reference." sentence rather than leave a
permanent 404; the surrounding prose still points at the Software
Guide and CONTRIBUTING.md as starting points.

hjmjohnson committed:
DOC: Modernize Insight Journal URLs in 3.2 release notes

Insight Journal moved from the legacy
   InsightJournalManager/view_reviews.php?...&pubid=N
URL scheme to the canonical
   /browse/publication/N
form years ago.  All 14 publication URLs flagged by
Utilities/Maintenance/check-links.sh in
Documentation/docs/releases/3.2.md are mechanically rewritten to the
modern path; each was verified to return HTTP 200 individually.

hjmjohnson committed:
DOC: Repoint legacy itk.org/Wiki links to ITKWikiArchive

The itk.org wiki was retired and snapshot-archived under
insightsoftwareconsortium.github.io/ITKWikiArchive (gh-pages source at
github.com/InsightSoftwareConsortium/ITKWikiArchive).  Four entries
in the 4.0 release notes still pointed at the dead itk.org/Wiki/...
URLs (or used a malformed escape sequence on the archive URL).
Update them to the verified archive paths:

- Modern_C\%2B\%2B (broken backslash escape)
    -> Modern_C%252B%252B/ (the directory's own name is double-encoded
       in the archive layout).
- itk.org/Wiki/Refactoring_itk::FEM_framework_-_V4
    -> ITK_Release_4/Refactoring_FEM_Framework/.
- itk.org/Wiki/Refactoring_Level-Set_framework_-_V4
    -> ITK/Release_4/Refactoring_Level_Set_Framework/Refactoring_Level_Set_Framework/.
- itk.org/Wiki/GPU_Acceleration_-_V4
    -> ITK_Release_4/GPU_Acceleration/GPU_Acceleration/.

Each replacement was verified to return HTTP 200 individually.

hjmjohnson committed:
DOC: Modernize 1.0 release-note URLs to current Kitware/ITK successors

The 1.0 release notes pre-date most of ITK's current infrastructure;
the broken URLs flagged by Utilities/Maintenance/check-links.sh have
well-defined modern successors:

- public.kitware.com/dashboard.php?name=itk (the legacy Kitware
  dashboard) and public.kitware.com/Dart (Dart, the predecessor of
  CDash) -> open.cdash.org/index.php?project=Insight.
- public.kitware.com/Cable (the CABLE C++ wrapping system) ->
  github.com/CastXML/CastXML.  CABLE was succeeded by GCC-XML and
  then by CastXML, which is the wrapping toolchain ITK uses today.
- www.cmake.org/CMake/HTML/Download.html -> cmake.org/download/.
- www.itk.org/HTML/Download.php -> docs.itk.org/en/latest/download.html.
- www.itk.org/HTML/Examples.htm -> examples.itk.org/.
- www.itk.org/mailman/listinfo/insight-users (the legacy mailing
  list) -> discourse.itk.org/ (the modern discussion forum that
  replaced it).

All replacements verified to return HTTP 200 individually.

hjmjohnson committed:
DOC: Repoint dead GDCM home in 2.0 release notes

creatis.insa-lyon.fr/Public/Gdcm/Main.html no longer resolves; GDCM
development moved to github.com/malaterre/GDCM (matches the same
fix in faq.md).

hjmjohnson committed:
DOC: Drop dead ITK Gerrit link from 4.0 release notes

review.source.kitware.com/p/ITK no longer resolves; the Gerrit
instance was decommissioned after the move to GitHub pull requests.
Replace the dead link with a parenthetical noting the migration.

hjmjohnson committed:
DOC: Drop the paywalled Scanco link from the faq formats table

The scanco.ch customer-login FAQ page is behind authentication and
returns 404 to anonymous fetches; the format-name cell stands on its
own without the link.  No public Scanco format reference is currently
discoverable to substitute.

hjmjohnson committed:
DOC: Repoint 5.0a01 Software Guide style links to current Book1 ch9

Both 'changes in style' and 'Coding Style Guide' references in
5.0a01.md pointed at Book1ch13.html#x57-259000C in the ITK Software
Guide.  The 5.x edition of the SG renumbered the coding-style chapter
to Book1ch9, and the per-section anchor IDs (`x57-...`) were
regenerated.  Drop the chapter-specific anchor and link to the
current chapter file (HTTP 200 verified); readers can navigate the
chapter TOC for the relevant section.

hjmjohnson committed:
DOC: Repoint or drop dead Kitware blog links in 4.8 release notes

Kitware retired the kitware.com/blog/home/post/N URL scheme; the
specific posts referenced in the 4.8 release notes were never
captured by the Wayback Machine and are not discoverable via the
modern Kitware site search.  Replace each broken "Details:" link
with the canonical ITK successor where one exists, and drop the
link entirely where no successor is available (the surrounding
prose already describes the topic):

- post/888 (CastXML wrapping replaces GCCXML)
    -> https://github.com/CastXML/CastXML
- post/912 (Emscripten / JavaScript build)
    -> https://wasm.itk.org/ (the canonical itk-wasm successor)
- post/890 (Software Guide HTML edition)
    -> https://itk.org/ITKSoftwareGuide/html/
- post/904 (cross-compilation/packaging),
  post/887 (Raspberry Pi),
  post/893 (Android),
  post/883 (MXE/MinGW-w64),
  post/891 (POWER8),
  post/899 (UpdateThirdPartyFromUpstream.sh / Git subtree)
    -> "Details:" link removed; topic line retained.

hjmjohnson committed:
DOC: Repoint or drop dead Kitware blog links in 4.9 release notes

Same Kitware-blog URL retirement as the 4.8 fixup; canonical
successors used where available, dead link dropped otherwise:

- post/942 (AnisotropicDiffusionLBR web-browser reproducibility)
    -> https://wasm.itk.org/ (the itk-wasm in-browser runtime that
       evolved out of the original Emscripten experiments).
- post/997 (External Modules outside the ITK source tree)
    -> https://docs.itk.org/en/latest/contributing/module_workflows.html
       (the canonical module workflows doc).
- post/939 (Option to export all library symbols on Windows)
    -> "Details:" link removed; topic line retained.

hjmjohnson committed:
DOC: Repoint dead Insight Journal handles to canonical IJ publication URLs

Two hdl.handle.net handles consistently return HTTP 500; the matching
publications were located in the modern insight-journal.org catalog
and verified to return HTTP 200:

- 10380/320 (SplitComponents new-class entry in 4.6 release notes)
    -> insight-journal.org/browse/publication/774
       ("An ITK Class that Splits Multi-Component Images").
- 1926/3596 (SLIC super-pixel filter in 5.0b01 release notes,
  referenced twice in the same file)
    -> insight-journal.org/browse/publication/989
       ("Scalable Simple Linear Iterative Clustering (SSLIC) Using a
       Generic and Parallel Approach", Lowekamp et al.).

Each replacement was confirmed by matching publication metadata in
the IJ /browse listing.

hjmjohnson committed:
DOC: Note that MIND DICOM-tool distribution URLs are no longer reachable

Both jeffro.net/mind and caddlab.rad.unc.edu/software/MIND have been
offline for years and have no Wayback Machine snapshot.  Drop the
hyperlinks, keep the URLs as inline code so the historical record of
where the project was hosted is preserved, and add a one-clause note
that the URLs are no longer reachable.

hjmjohnson committed:
DOC: Repoint NLM ITKv4 kick-off agenda to its Wayback snapshot

visual.nlm.nih.gov no longer hosts the 2010 ITKv4 kick-off meeting
agenda; the Internet Archive captured the page on 2012-03-13.
Repoint the link to the Wayback snapshot so the historical
information remains reachable.

hjmjohnson committed:
ENH: lychee.toml absorbs gateway-timeout and bot-block status codes

After all link-rot fixes for ITK's documentation, the periodic
check-links.sh run still surfaced ~14 "errors" that were not link rot
but transient infrastructure responses from the maintainer's network:

- 504 from eth.limo gateway in front of content-link-upload.itk.eth.limo.
- 522 from Cloudflare in front of opencollective.org.
- TCP-level resets and connection-failed signals from ipfs.io,
  monai.io, dicom.nema.org.

Extend the accept list to cover 504 and 522, and turn on
accept_timeouts so the residual reachability artifacts don't drown the
report.  Real link rot (404, 5xx other than 504/522, etc.) still
surfaces normally.

hjmjohnson commented on Utilities/Maintenance/check-links.sh:
```diff
@@ -0,0 +1,96 @@
+#                        https://github.com/lycheeverse/lychee/releases
+#   - bash >= 4
+
+set -euo pipefail
```
Fixed in 30c4962540 — dropped `-e` from the `set` line so lychee's non-zero exit on findings still allows `status=$?` and the 'Report written to ...' echo to run.

hjmjohnson commented on Utilities/Maintenance/lychee.toml:
```diff
@@ -0,0 +1,61 @@
+
+exclude_path = [
+    "Modules/ThirdParty",
+    "Documentation/Release",  # release-notes commit-URL false positives
```
Fixed in 30c4962540 — dropped the stale `Documentation/Release` exclude_path entry entirely.  Scanning the release notes is the desired behaviour now that we have a commit-URL exclude regex covering the original false-positive case.

hjmjohnson commented on Documentation/docs/releases/1.0.md:
```diff
@@ -77,24 +77,24 @@ Insight/Documentation/InsightStart.pdf is highly recommended.
+  + https://open.cdash.org/index.php?project=Insight - the ITK software
     quality dashboard
-  + http://public.kitware.com/Dart - what manages the testing process in ITK
-  + http://public.kitware.com/Cable - the automated C++ wrapping system
+  + https://open.cdash.org/index.php?project=Insight - what manages the testing process in ITK
+  + https://github.com/CastXML/CastXML - the automated C++ wrapping system
```
Fixed in 15f15d8eed — collapsed the duplicate CDash row into a single bullet that mentions the dashboard also manages the testing process.

jhlegarreta commented:
> Holler if you'd rather I close this and open the script under a fresh PR — both work; reusing this branch felt cleaner since the conversation history is already here.

@hjmjohnson This is fine. Thanks for all this work.

hjmjohnson commented on Documentation/docs/migration_guides/itk_5_migration_guide.md:
```diff
@@ -624,14 +624,6 @@ To resolve these errors, update the remote module's dependency specification in
 `itk-module.cmake`. List the dependent module under `DEPENDS` instead of
 `COMPILE_DEPENDS`; this will link to the new dependent library.
 
-Update scripts
```
This should be preserved with a permanent link based on the hash that has the link, instead of the path in the 'main' directory.

hjmjohnson replied:
Addressed in ec235d8 — restored the `Update scripts` section with a permalink to commit `21851448ad` (the last commit where `Utilities/ITKv5Preparation/` still existed) so the scripts remain accessible to anyone bringing an old ITKv4 codebase forward.

hjmjohnson commented on Documentation/docs/learn/courses.md:
```diff
@@ -7,18 +7,15 @@ Kitware also offers periodic [short online courses given by expert instructors](
 - The University of Iowa-Medical [Image Analysis course](https://data.kitware.com/#collection/57b5c9e58d777f126827f5a1/folder/57b5d18b8d777f10f269311c) is semester-long course taught by Hans Johnson. It is an excellent introduction to medical image analysis and ITK. The course covers the architecture of ITK, programming with ITK, medical image representation, filtering, affine registration, and d ...
 - A long-running [Methods in Medical Image Analysis course](https://www.cs.cmu.edu/~galeotti/methods_course/) from John Galeotti at Carnegie-Mellon has both lecture Powerpoints and high quality video recordings of the lectures available. Its purpose, from its syllabus, is “to gain theoretical and practical skills in medical image analysis, including skills relevant to general image analysis.  The ...
 - The University of Pennsylvania, Center for Biomedical Image Computing and Analytics has made available the presentations and code from a on "[CPP for Image Processing and Machine Learning](https://www.nitrc.org/projects/cbica_tutorials)."  The tutorial intends to help someone with a good theoretical background in image processing and machine learning quickly prototype algorithms using C++ and e ...
-- A [Medical Image Computing course](http://www.cs.ucf.edu/~bagci/teaching/mic17.html) at the University of Central Florida covers the basics of radiological image modalities and their clinical use, an introduction to medical image computing and toolkits, image filtering, enhancement, noise reduction, and signal processing, medical image registration, medical image segmentation, medical image vis ...
```
Course description should stay; only the link should be removed.

hjmjohnson replied:
Addressed in 038b3c7 — restored the course bullet with its full description; only the broken link itself was dropped, with a trailing note pointing at the (now-unreachable) original URL.

hjmjohnson commented on Documentation/docs/learn/courses.md:
```diff
@@ -7,18 +7,15 @@ Kitware also offers periodic [short online courses given by expert instructors](
 - The University of Pennsylvania, Center for Biomedical Image Computing and Analytics has made available the presentations and code from a on "[CPP for Image Processing and Machine Learning](https://www.nitrc.org/projects/cbica_tutorials)."  The tutorial intends to help someone with a good theoretical background in image processing and machine learning quickly prototype algorithms using C++ and e ...
-- A [Medical Image Computing course](http://www.cs.ucf.edu/~bagci/teaching/mic17.html) at the University of Central Florida covers the basics of radiological image modalities and their clinical use, an introduction to medical image computing and toolkits, image filtering, enhancement, noise reduction, and signal processing, medical image registration, medical image segmentation, medical image vis ...
 - An [Image Registration course](https://www.cs.rpi.edu/courses/spring04/imagereg/) at the Rensselaer Polytechnic Institute (RPI) covers all aspects of image registration, including the underlying mathematics, the images themselves, algorithms, implementations, and applications. Applications range from medical imaging to automatic geometric modeling. Special emphasis is given to software toolkits ...
-- The [Computer Assisted Image Analysis course](https://www.it.uu.se/edu/course/homepage/bild1/vt14) at Uppsala University gives an introduction to the processing and analysis of digital images, covering: pointwise image operators, local image operators, Fourier analysis of images, mathematical morphology and distance transforms, image segmentation, object description, classification, color image ...
```
Course description should stay; only the link should be removed.

hjmjohnson replied:
Addressed in 038b3c7 — restored the course bullet with its full description; only the broken link itself was dropped, with a trailing note pointing at the (now-unreachable) original URL.

hjmjohnson commented on Documentation/docs/learn/courses.md:
```diff
@@ -7,18 +7,15 @@ Kitware also offers periodic [short online courses given by expert instructors](
-- A [Medical Image Computing course](http://www.cs.ucf.edu/~bagci/teaching/mic17.html) at the University of Central Florida covers the basics of radiological image modalities and their clinical use, an introduction to medical image computing and toolkits, image filtering, enhancement, noise reduction, and signal processing, medical image registration, medical image segmentation, medical image vis ...
 - An [Image Registration course](https://www.cs.rpi.edu/courses/spring04/imagereg/) at the Rensselaer Polytechnic Institute (RPI) covers all aspects of image registration, including the underlying mathematics, the images themselves, algorithms, implementations, and applications. Applications range from medical imaging to automatic geometric modeling. Special emphasis is given to software toolkits ...
-- The [Computer Assisted Image Analysis course](https://www.it.uu.se/edu/course/homepage/bild1/vt14) at Uppsala University gives an introduction to the processing and analysis of digital images, covering: pointwise image operators, local image operators, Fourier analysis of images, mathematical morphology and distance transforms, image segmentation, object description, classification, color image ...
-- An [Advanced Image Processing and Analysis course](https://www.eng.uwo.ca/biomed/courses/courses_9519.html) at Western University explores a few major areas of digital image processing at an advanced level, with primary emphasis on medical applications. Topics covered include image segmentation, image registration, validation of image processing algorithms, and image processing using 3D Slicer  ...
```
Course description should stay; only the link should be removed.

hjmjohnson replied:
Addressed in 038b3c7 — restored the course bullet with its full description; only the broken link itself was dropped, with a trailing note pointing at the (now-unreachable) original URL.

hjmjohnson commented on Documentation/docs/learn/faq.md:
```diff
@@ -40,21 +40,21 @@ images.
 | [MetaImage](https://insightsoftwareconsortium.github.io/ITKWikiArchive/Wiki/ITK/MetaIO/Documentation) (.mha/.mhd) | [`itk::MetaImageIO`](https://docs.itk.org/projects/doxygen/en/stable/classitk_1_1MetaImageIO.html) | |
 | [MINC 2.0](https://en.wikibooks.org/wiki/MINC/SoftwareDevelopment/MINC2.0_File_Format_Reference) (Medical Image NetCDF) | [`itk::MINCImagIO`](https://docs.itk.org/projects/doxygen/en/stable/classitk_1_1MINCImageIO.html) | |
-| [MGH](https://surfer.nmr.mgh.harvard.edu/fswiki/FsTutorial/MghFormat) | [`itk:MGHImageIO`](https://docs.itk.org/projects/doxygen/en/stable/classitk_1_1MGHImageIO.html) | |
+| [MGH](https://surfer.nmr.mgh.harvard.edu/fswiki/FsTutorial/MghFormat) | [`itk::MGHImageIO`](https://github.com/InsightSoftwareConsortium/ITK/blob/main/Modules/IO/MGHIO/include/itkMGHImageIO.h) | |
```
use permanent link.

hjmjohnson replied:
Addressed in f2a2288 — replaced the `/blob/main/` link with the matching doxygen permalink (`docs.itk.org/projects/doxygen/en/stable/classitk_1_1MGHImageIO.html`), which is what every other ImageIO row uses.

hjmjohnson commented on Documentation/docs/releases/5.0b01.md:
```diff
@@ -72,22 +72,22 @@ Based on experiences since initial changes introduced in [ITK 5 Alpha 2](https:/
 - `SetNumberOfThreads` has been renamed to `SetNumberOfWorkUnits` to more accurately reflect configuration of work unit granularity, especially in the context of the Threading Building Blocks (TBB) backend.
 - Progress reporting with multiple parallel operations is enabled with a new `itk::ProgressTransformer` class.
-- A large corpus of code has been transitioned ITK 5, and the process is now relatively straight forward and well documented. Please see the [ITK 5 Migration Guide](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/ITK5MigrationGuide.md) for more information.
+- A large corpus of code has been transitioned ITK 5, and the process is now relatively straight forward and well documented. Please see the [ITK 5 Migration Guide](https://github.com/InsightSoftwareConsortium/ITK/blob/main/Documentation/docs/migration_guides/itk_5_migration_guide.md) for more information.
```
use permanent link instead of `main` reference

hjmjohnson replied:
Addressed in 7cc2c70 — replaced the `/blob/main/` migration-guide ref with a SHA-pinned permalink (`/blob/59176b1dca/...`) so the link stays valid as `main` advances.

hjmjohnson reviewed (approved):
Thanks @jhlegarreta for initiating this effort.
